### ISSUE TITLE
Nested class definition does not satisfy a ClassVar protocol member (new 0.0.70 bug)

### ISSUE THREAD

**Issue by jamtat** [astral-sh/ty#4234]
### Summary

As of ty 0.0.70 nested class definitions no longer satisfy a protocol requiring a `type` as a `ClassVar` member. 

[I have created an example repo here](https://github.com/jamtat/ty-classvar-repro) and also included a minimal reproduction of the issue below. Prior to `0.0.70` `Nested` satisfied ty as an implementor of the `HasKey` protocol.

```python
from typing import Any, ClassVar, Protocol


class HasKey(Protocol):
    Key: ClassVar[type[Any]]


# A nested class definition no longer satisfies the protocol
# as of ty==0.0.70
class Nested:
    class Key:
        pass


# Error observed:
# Object of type `Nested` is not assignable to `HasKey`: Incompatible value of type `Nested`
# info: type `Nested` is not assignable to protocol `HasKey`
# info: └── protocol member `Key` is incompatible
# info:     └── protocol member `Key` is an instance variable on type `Nested`, but a class variable is requiredtyinvalid-assignment
a: HasKey = Nested()


# Accepted; the same class bound by an assignment
class _Key:
    pass


class Accepted:
    Key: ClassVar[type[Any]] = _Key


x: HasKey = Accepted()
```

### Version

0.0.70

**Comment by carljm** [comment]
Thanks for the report, and sorry for the regression!

This change was intentional, in the sense that we intended to require `ClassVar` to satisfy a protocol `ClassVar`. This new behavior matches both mypy and pyright (including in the case of a nested class - they both emit the same error we do.)

That said, it seems to me that a nested class probably _should_ be considered as a `ClassVar` implicitly -- it definitely doesn't define an instance variable. So I would be in favor of special-casing nested classes here. @AlexWaygood, does that make sense to you?

**Comment by jamtat** [comment]
```
In [1]: class A:
   ...:     class B:
   ...:         pass
   ...:

In [2]: a = A()

In [3]: a.B
Out[3]: __main__.A.B

In [4]: class C:
   ...:     pass
   ...:

In [5]: A.B = C

In [6]: a.B
Out[6]: __main__.C
```

Python seems to treat internal class declarations as ClassVars - at face value it seems like mypy and pyright are wrong here unless there's some extra special machinery I'm not familiar with.

This is a pretty easy one to work around so it's not critical - I just so happen to be working on a large codebase that is acting as a good testing ground for your work. ty has made a huge difference for us as we can now run type checking on pre-commit, pylance took minutes on the whole repo.

**Comment by AlexWaygood** [comment]
Hey @jamtat, thanks for the report!

Our behaviour changed here as a result of https://github.com/astral-sh/ruff/pull/27530, which was aimed at fixing some assertions in the conformance suite. We're a bit undecided about whether we'll keep the behaviour that that PR introduced -- it seems to me that the conformance-suite assertions aren't very well grounded in the language of the typing spec here, so I don't have full confidence that they're asserting the correct thing.

That said, even if we do go back to the behaviour we had before https://github.com/astral-sh/ruff/pull/27530, I'm not _totally_ sure your code should pass. In order for an attribute to be a `ClassVar[type[Any]]`, reading that attribute should result in a type assignable to `type[Any]`, and it should be valid to write to the attribute with a value of `type[Any]`. The read requirement is fulfilled here (`<class 'Nested'>` is assignable to `type[Any]`), but the write requirement is not: we would generally not allow you to do `Nested.Key = <some other class>`. If we did, it wouldn't be sound to infer the type of `Nested.Key` as `<class 'Nested.Key'>`; we'd have to infer it with a much less precise type like `type[object]`, and that could cause false positives elsewhere.

It's possible a `@property` protocol member might be a better fit for you here, since in order for an attribute to satisfy a `@property` member it only needs to be readable (writability is not required)

**Comment by jamtat** [comment]
Hey @AlexWaygood thanks for the detailed explanation, that makes sense to me.

The following declaration works and passes `ty` without having to pull the "Key" type decl out. Good pointer on the property being readonly as well. Thanks.

```python
class HasKey(Protocol):
    @property
    def Key(self) -> type[Any]: ...

class Nested:
    class Key:
        pass

a: HasKey = Nested()

# ty reports:
# Cannot assign to attribute `Key` on type `HasKey`
cast(HasKey, a.Key).Key = Nested.Key
```

**Comment by carljm** [comment]
@AlexWaygood would it make sense to use this issue to track the possibility of rolling back https://github.com/astral-sh/ruff/pull/27530, or do you find it clearer to close this as "not planned" (since the OP code here should not pass checking either way) and file a new issue for reconsidering https://github.com/astral-sh/ruff/pull/27530? (Either way, I'll let you take care of re-titling this issue or filing the new one.)

EDIT: on second thought, I think we should close this, so I'll do that. You can file a new issue for the broader question.
